azure-pipelines.yml

```
@@ -0,0 +1,114 @@
+# Azure DevOps pipeline to build, check source codes, run tests, and more.
+
+variables:
+    YARN_CACHE_FOLDER: $(Pipeline.Workspace)/.yarn
+  
+  pool:
+    vmImage: 'ubuntu-latest'
```

```suggestion
    vmImage: 'windows-2019'
```
since we are deploying on azure functions windows runtime, probably it's better to use a windows vm

---

@gunzip I tried to run the pipeline with `vmImage: 'windows-2019'` but several syntax errors occurred. 
For instance, in the _Build_ step of the _compile_ job (stage: _Build_check_), this is the result of the execution of the `yarn build` command:   
```
Starting: Build
==============================================================================
Task         : Command line
Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
Version      : 2.163.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
==============================================================================
Generating script.
Script contents:
yarn build
========================== Starting Command Output ===========================
"C:\windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "d:\a\_temp\79a3529b-5427-4b37-80f3-c82ead3bf936.cmd""
yarn run v1.21.1
$ yarn generate:definitions && tsc
$ rimraf ./generated/definitions && mkdir -p ./generated/definitions && gen-api-models --api-spec ./openapi/index.yaml --out-dir ./generated/definitions
The syntax of the command is incorrect.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
##[error]Cmd.exe exited with code '1'.
Finishing: Build
```

Do you have any suggestions?

---

can you try using "bash" instead of "script" ?
ie. 

```yaml
steps:
- bash: |
    yarn ...

```

---

I already tried using `bash` but without success! It's not enough...

After many attempts, I have found the root cause of the problems... In the `packege.json` file there is the definition of some scripts used in the pipeline:
```
{
  "name": "io-functions-admin",
  "version": "0.1.0",
  "license": "MIT",
  "scripts": {
    "build": "yarn generate:definitions && tsc",
    "build:production": "npm run prestart && npm prune --production",
    "watch": "tsc --w",
    "prestart": "npm run build && func extensions install",
    "start:host": "func start",
    "start": "npm-run-all --parallel start:host watch",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "lint": "tslint -p .",
    "deploy": "npm run build && func azure functionapp publish agid-functions-admin-test",
    "generate:definitions": "rimraf ./generated/definitions && mkdir -p ./generated/definitions && gen-api-models --api-spec ./openapi/index.yaml --out-dir ./generated/definitions"
  },
```
```
          - bash: |
              yarn build
            displayName: 'Build'
```
The `yarn build` command fails when running in a Windows bash for the following reasons: 
- `mkdir -p` is not supported 
- `&&` can be used only one time (i.e. _cmd1 && cmd2_ is ok but _cmd1 && cmd2 && cmd3_ fails)

So, it seems that there are some limitations when executing a bash script on Azure pipelines... I could easily solve the problem avoiding re-using the scripts defined in the `packege.json`, for instance changing the above job in the pipeline with the following:
```
        - bash: |
            mkdir ./generated && mkdir ./generated/definitions 
            yarn gen-api-models --api-spec ./openapi/index.yaml --out-dir ./generated/definitions 
            yarn tsc
          displayName: 'Build'
``` 
In this way, the pipeline runs without problems both in 'ubuntu-latest' and 'windows-2019' and we only need to take into account that developers could use little different scripts to build the code in their local environments. 

I also notice that the pipeline execution time is different:
- vmImage: 'ubuntu-latest'    takes about 50s
- vmImage: 'windows-2019'  takes about 1m 20s

If all the above limitations are reasonable I can proceed with the specified changes to run the pipeline in Windows vm. 
@gunzip what do you think?

---

Thank you for the insight @raicastino ! IMHO it's a duty of the package owners to maintain compatibility. As we run everything inside a windows VM we must make package.json tasks cross platform.

---

@gunzip Following your suggestions:
- I changed vmImage from 'ubuntu-latest' to 'windows-2019' 
- Changed the package.json file to make a script cross-platform (using shx)
- For the build job I used in the pipeline the script defined in the package.json (yarn build)

---

great ! can we test a deploy into the staging environment ? (how ? do we have to create an azure pipeline service connection ? in which account ?)

---

Actually, this pipeline will be a replace of the current pipeline based on CircleCI. I need permissions on GitHub to enable this pipeline. By the way, only the CI part is currently implemented...